pandas/core/indexes/category.py

```diff
@@ -277,6 +278,26 @@ def _is_dtype_compat(self, other) -> Categorical:
 
         return other
 
+    @doc(Index.astype)
+    def astype(self, dtype, copy: bool = True) -> Index:
```

Comment:
can you comment on why you're doing this at this level?

Reply:
This concerns casting to/from NumIndexes and CategoricalIndexes. Doing e.g. `num_index.astype("category").astype("int32")` should return a `NumIndex[int32]`. Doing it similarly with `Int64Index` should similary return a `Int64Index`  etc.

Reply:
Added a comment explaining. this

Reply:
Not really, `Categorical.astype` returns a numpy array and when converting that back into an index ,we can't know if that should be put back into a `NumericIndex` or a `[Int|UInt|Float]64Index`.

All this is however just done in order to keep backward compat with the old numeric index types. If/when we drop the old numeric index types in pandas 2.0, this kind of thing can all be removed, because there will only be one numeric index class.

In general everything using `_is_numeric_index`, it can be removed when we drop the old numeric index types, so this will all simplify when that happens.

Reply:
Alternatively, we could also for now not have the full index -> categorical -> index roundtrip faithfully keep the NumericIndex class? (that seems a relatively minor corner case?)

Reply:
I replied below:

> it simplifies the tests suite quite a bit returing the same index class as categories, because assert_index_equal would complain if the result from astype would bw different than for categories.

You commented:

> OK, that sounds as a good reason to indeed already add this special case in astype.

So I'll mark this resolved.

Reply:
yeah i am puzzled why this is done at this level. can you comment here (and not resolve)